torch/tensor.py

```diff
@@ -441,4 +441,52 @@ def __array_wrap__(self, array):
             array = array.astype('uint8')
         return torch.from_numpy(array)
 
+    @property
+    def __cuda_array_interface__(self):
+        """Array view description for cuda tensors.
+
+        See:
+        https://numba.pydata.org/numba-doc/latest/cuda/cuda_array_interface.html
+        """
+
+        # raise AttributeError for unsupported tensors, so that
+        # hasattr(cpu_tensor, "__cuda_array_interface__") is False.
+        if not self.device.type == "cuda":
+            raise AttributeError(
+                "Not on cuda device, use Tensor.cuda() first: %r" %
+                self.device
+            )
+
+        if self.is_sparse:
+            raise AttributeError(
+                "Can't convert sparse tensor, use Tensor.to_dense() "
+                "to convert to a dense tensor first."
+            )
+
+        # RuntimeError, matching tensor.__array__() behavior.
+        if self.requires_grad:
+            raise RuntimeError(
+                "Can't get __cuda_array_interface__ on Variable that requires grad. "
+                "Use var.detach().__cuda_array_interface__ instead."
+            )
+
+        typestr = {
```

Comment:
Aren't you obligated to specify endianness in the type string?

Reply:
This is currently working because numba assumes the native byte order of the device if it isn't specified, though I agree that this (may) a bug in my implementation.

I'm not super familiar with the `pytorch` internals and don't have access to a big-endian system, so please feel free to correct me, but I believe:

* `pytorch` currently does not support non-native byte orders when [converting from numpy](https://github.com/asford/pytorch/blob/e7c7a45c620cd3d8f76e3f3906e557d7737d0c12/torch/csrc/utils/tensor_numpy.cpp#L124-L128).
* `pytorch` does not have a way of programmatically accessing the byteorder of tensor structures, which are all represented in native ordering.
* Numba converts dtype representations via [`numpy.dtype`](https://github.com/numba/numba/blob/master/numba/cuda/api.py#L194-L204), which will use the host-endianness if explicitly native ("=") or not specified (as in this implementation).
* Cuda devices are little-endian, as are `x86` and `ppc64le` deployments.

I think the options are:

1. Specify all little-endian via "<".
2. Specify 'native' byte ordering via "=", which will currently break if there was some form of `ppc64`/`cuda` cross deployment but _might_ more accurately reflect the semantics of the underlying buffer.

I will update this to return "<", unless there's a missed assumption above.

Reply:
I'm definitely not saying that anyone is actually using big-endian, but it's simply a matter of hewing to the standard. "<" seems fine.